crates/pathfinder/src/sequencer/reply.rs

```diff
@@ -237,41 +237,74 @@ pub mod transaction {
+        #[serde_as(as = "FeeAsHexStr")]
+        pub max_fee: Fee,
+        pub nonce: TransactionNonce,
+        pub sender_address: ContractAddress,
+        #[serde_as(as = "Vec<TransactionSignatureElemAsDecimalStr>")]
         #[serde(default)]
-        pub calldata: Option<Vec<CallParam>>,
-        /// None for Invoke, Some() for Deploy and Declare
+        pub signature: Vec<TransactionSignatureElem>,
+        pub transaction_hash: StarknetTransactionHash,
+        pub r#type: Type,
+        #[serde_as(as = "TransactionVersionAsHexStr")]
+        pub version: TransactionVersion,
+    }
+
+    /// Represents deserialized L2 deploy transaction data.
+    #[serde_as]
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    #[serde(deny_unknown_fields)]
+    pub struct DeployTransaction {
+        pub contract_address: ContractAddress,
+        pub contract_address_salt: ContractAddressSalt,
         #[serde(default)]
         pub class_hash: Option<ClassHash>,
-        #[serde_as(as = "Option<Vec<ConstructorParamAsDecimalStr>>")]
-        #[serde(default)]
-        pub constructor_calldata: Option<Vec<ConstructorParam>>,
-        /// None for Declare
-        #[serde(default)]
-        pub contract_address: Option<ContractAddress>,
-        #[serde(default)]
-        pub contract_address_salt: Option<ContractAddressSalt>,
-        #[serde(default)]
-        pub entry_point_type: Option<EntryPointType>,
-        #[serde(default)]
-        pub entry_point_selector: Option<EntryPoint>,
-        #[serde_as(as = "Option<FeeAsHexStr>")]
-        #[serde(default)]
-        pub max_fee: Option<Fee>,
-        #[serde(default)]
-        pub nonce: Option<TransactionNonce>,
-        /// Some() for Declare
-        #[serde(default)]
-        pub sender_address: Option<ContractAddress>,
-        #[serde_as(as = "Option<Vec<TransactionSignatureElemAsDecimalStr>>")]
-        #[serde(default)]
-        pub signature: Option<Vec<TransactionSignatureElem>>,
+        #[serde_as(as = "Vec<ConstructorParamAsDecimalStr>")]
+        pub constructor_calldata: Vec<ConstructorParam>,
+        pub transaction_hash: StarknetTransactionHash,
+        pub r#type: Type,
+    }
+
+    /// Represents deserialized L2 invoke transaction data.
+    #[serde_as]
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    #[serde(deny_unknown_fields)]
+    pub struct InvokeTransaction {
+        #[serde_as(as = "Vec<CallParamAsDecimalStr>")]
+        pub calldata: Vec<CallParam>,
+        pub contract_address: ContractAddress,
+        pub entry_point_selector: EntryPoint,
+        pub entry_point_type: EntryPointType,
+        #[serde_as(as = "FeeAsHexStr")]
+        pub max_fee: Fee,
+        #[serde_as(as = "Vec<TransactionSignatureElemAsDecimalStr>")]
+        pub signature: Vec<TransactionSignatureElem>,
         pub transaction_hash: StarknetTransactionHash,
         pub r#type: Type,
         #[serde_as(as = "Option<TransactionVersionAsHexStr>")]
```

Comment:
[I think this is actually a bug](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/starknet/services/api/feeder_gateway/response_objects.py#L284), can't recall why I put it here in the first place.